Speedometer refactor

braem opened the pull request:
Closes #638 & closes #304

- Created a `SpeedometerLabel` class that handles
    - Setting label text
    - Unit conversion
    - Colorization
    - Fadeout
- Labels simply call `Update` with the speed value. Done in `OnThink()` or `FireGameEvent()` depending on the label.
    - Abs/horiz speedos call it in `OnThink()`
    - Last jump velocity & stage enter/exit call it in `FireGameEvent()`
        - Added an event that's fired when the player jumps to do this
- Better defined update types
    1. No color
    2. Range-based (with ranges being defined by user)
    3. Comparison-based
    4. Comparison-based but comparison is written to a separate label (a la stage enter/exit)
        - Other labels can use this, like last jump velocity. Apparently bhoppers have been wanting this
- Added a user-defined speedometer label order
- Created a VDF file for speedometer customization
    - Each speedometer can have their own units, colorization mode, and colorization ranges
        - Removed the units label as they can all be different. Might be worth adding a small one that's pinned to the upper right of each label.
    - Can define the speedometer label order
    - New con commands `mom_hud_speedometer_loadcfg` and `mom_hud_speedometer_savecfg` to load/save this file.
        - These take an optional parameter of the gamemode, so you can load the surf setup for rj, etc.
- Updated the hud settings for this, except for custom colorization ranges & ordering. These are better tackled in a different PR as this one is already quite big
    - Dropdowns for gamemode, speedo label type, units, and colorization. Checkbutton for on/off.
    - Gamemode and label type comboboxes load the gamemode / label, while the others write to the file.
- Removed unneeded cvars, since everything is loaded from file. Will update the docs page when this gets merged in

Apologies for the "Major speedometer refactor" commit, I wasn't able to separate out individual commits as nearly the entire file was re-written. I did manage to split the speedometer label into several commits though.

### Checklist
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

braem commented:
On second thought, that VDF file is quite big. What are your thoughts on having separate VDF files for each gamemode in a subfolder (ie. cfg/speedometer/rj_default.vdf) ? Would avoid the overhead of writing the entire file every time.

Also, is autosaving when changing in hud settings fine? I dont experience any lag with it but then again I'm on a fast NVMe SSD. Could move the saving to only on settings/game close or tab change.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,248 @@
+bool SpeedometerLabel::StartFadeout()
+{
+    Assert(HasFadeOutAnimation());
+    g_pClientMode->GetViewportAnimationController()->StopAnimationSequence(GetParent(), m_pSzAnimationName);
+    *m_pFlAlpha = 255.0f;
+    return g_pClientMode->GetViewportAnimationController()->StartAnimationSequence(m_pSzAnimationName);
```
I've been thinking about this a lot, and this is actually an underlying issue with animations. To run one from HudAnimations.txt requires the name of the panel so that it knows what to do, but that seems silly, because the only other alternative is to *directly run animations on a panel*. I'm thinking there can be a solution in AnimationController that uses a predefined animation, but runs it for the given panel, like:

`StartAnimationSequenceForPanel( Panel *pPanel, const char *pSequenceName )`

And this would just basically run the sequence in the context of the panel calling it. In this case it would be our speedometer labels, so then the sequence would define the thing it's animating (the panel's alpha, for example) and have it animate the property it should for that control.

Unfortunately this means the animations would have to be very specific to specific panels and contexts, but does allow user customization and easy tweaking in the long run.

braem replied:
That would be very useful here! I assume not in this PR though :p

Gocnak replied:
Yeah I'll tackle this in my ramp speeds branch.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,248 @@
+                gravity;
+        }
+        break;
+    case SPEEDOMETER_UNITS_UPS:
```
Handled in the early return above.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,248 @@
+
+void SpeedometerLabel::Update(float value)
+{
+    if (m_flNextUpdateCheck > gpGlobals->curtime)
+        return;
```
Was there a performance reason to do this? Should this be a cvar?

braem replied:
Could be a cvar yeah. Not really a performance problem, just figured it didn't need to update every `OnThink`. Removing it messes with the colorization since the diff between current and past values varies depending on framerate.  Before it was only setting the past value when color and using `MOM_COLORIZATION_CHECK_FREQUENCY`. 

I just realized this when looking back at the previous code. I'll do that instead.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,248 @@
+    ConvertUnits();
+    SetText(m_flCurrentValue);
+    Colorize();
+    m_bDoneFading = HasFadeOutAnimation() ? !StartFadeout() : false;
```
Why is this done here as well? Isn't this updated in the `OnThink` method above?

braem replied:
This stops and starts the fadeout when the label is updated, while `OnThink` handles constantly setting the alpha of it & resetting it's values when it's done fading out.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,248 @@
+    HFont labelFont = GetFont();
+    int combinedLength = UTIL_ComputeStringWidth(labelFont, szMain);
+
+    if (m_eColorizeType == SPEEDOMETER_COLORIZE_COMPARISON_SEPARATE && m_bDrawComparison)
```
Are the `COMPARISON_SEPARATE` colorize type and `m_bDrawComparison` bools glued together? Like is there a case of where you want that colorization style but don't want the comparison label to draw?

braem replied:
Yes, when the stage enter/exit velocity doesn't have a comparison to draw.

braem commented:
Pushed a few changes to resolve some of the more minor issues, mainly to get them out of the way since there are so many. I will continue to work on the rest.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.h:
```diff
@@ -0,0 +1,98 @@
+#include <vgui_controls/Label.h>
+#include "mom_player_shared.h"
+#include "mom_shareddefs.h"
+#include "utlvector.h"
```
I did this with `IScheme`, but was unable to do it with `C_MomentumPlayer` for some reason. It was giving errors whenever I removed the `mom_player_shared.h` include.

Maybe to do with it calling a static method? I've never forward declared a class in this manner before so I'm not sure.

Gocnak replied:
What kind of errors?

braem replied:
Just an "Incomplete type is not allowed" compiler error. I believe it is because the `C_MomentumPlayer::GetLocalMomPlayer()` function we are using returns a static variable in that file, so the file needs to be imported in order to be able to call it.

Gocnak replied:
Yeah, but why does the header need the import? Why not the cpp file?

braem replied:
Oh right, I moved that import to the source file now.

braem commented on mp/game/momentum/resource/momentum_english_ref.res:
```diff
@@ -187,6 +188,15 @@
+        "MOM_GameTypeID_RJ" "rj"
+        "MOM_GameTypeID_SJ" "sj"
+        "MOM_GameTypeID_Tricksurf" "tricksurf"
+        "MOM_GameTypeID_Ahop" "ahop"
```
Used these in the file, which trimmed down the console commands and loading/saving functions.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -28,150 +29,106 @@
+                g_pSpeedometer->SaveGamemodeData(g_pGameModeSystem->GetGameMode(i + 1)->GetType());
+                return;
+            }
+        }
```
Changed these to just have a number as yeah they aren't very user-facing.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -28,150 +29,106 @@
+        {
+            if (!Q_strcasecmp(args[1], g_pVGuiLocalize->FindAsUTF8(g_szGameModeIdentifiers[i])))
+            {
+                g_pSpeedometer->LoadGamemodeData(g_pGameModeSystem->GetGameMode(i + 1)->GetType());
```
Created a separate class to handle the file stuffs.

braem commented on mp/src/game/client/momentum/ui/SettingsPanel/HudSettingsPage.cpp:
```diff
@@ -77,18 +82,66 @@ HudSettingsPage::HudSettingsPage(Panel *pParent) : BaseClass(pParent, "HudSettin
+
+SpeedometerLabel* HudSettingsPage::GetSpeedoLabelFromType()
+{
+    return g_pSpeedometer->GetLabelList()[m_pSpeedometerType->GetCurrentItem()];
```
True, this array shouldn't be changed outside of the speedometer panel.

braem commented:
Implemented the requested changes, though need input on 4 of them.

Roughly, I
- Moved the file load/save stuffs to it's own class
- Removed those gametype string identifiers. As a result, let the introduced con commands just take in a number & changed the VDF file to use `#MOM_GameType_Surf`, etc.
- Cleaned up includes
- Moved colorizations into their own private functions
- Limited the colorization frequency instead of the entire update frequencying using `MOM_COLORIZATION_CHECK_FREQUENCY` as was done before 
- Moved all the options out of the constructor and into getters/setters
- Prevented users from setting the separate comparison colorization type for the abs/horiz speedos in the settings. They can still set this through editing the VDF file if they really want that.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.h:
```diff
@@ -0,0 +1,58 @@
+
+    SpeedoLabelList m_LabelOrderList;
+
+    SpeedometerLabel *m_Labels[SPEEDOMETER_MAX_LABELS] 
+        { m_pAbsSpeedoLabel, m_pHorizSpeedoLabel, m_pLastJumpVelLabel, m_pStageEnterExitVelLabel };
```
Let's keep constructing to the constructor.
```suggestion
    SpeedometerLabel *m_Labels[SPEEDOMETER_MAX_LABELS];
```

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,200 @@
+    if (!g_pFullFileSystem)
+        return;
+
+    KeyValues *pGamemodeKV = m_pGamemodeSetupData->FindKey(g_szGameModes[gametype]);
+    if (pGamemodeKV) // game mode is known
+    {
```
Early return if it's null.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+    SetPos(xOffset, GetYPos());
+}
+
+void SpeedometerLabel::OnThink()
+{
+    if (HasFadeOutAnimation() && !m_bDoneFading)
+    {
+        SetAlpha(*m_pFlAlpha);
+        m_pComparisonLabel->SetAlpha(*m_pFlAlpha);
+
+        if (CloseEnough(*m_pFlAlpha, 0.0f, FLT_EPSILON))
+        {
+            m_bDoneFading = true;
+            Reset();
+        }
+    }
+}
```
Always call base class for overrides. Early return for ease of reading.
```suggestion
void SpeedometerLabel::OnThink()
{
    BaseClass::OnThink();
    
    if (!HasFadeOutAnimation())
        return;
    
    if (m_bDoneFading)
        return;
    
    SetAlpha(*m_pFlAlpha);
    m_pComparisonLabel->SetAlpha(*m_pFlAlpha);
    if (CloseEnough(*m_pFlAlpha, 0.0f, FLT_EPSILON))
    {
        m_bDoneFading = true;
        Reset();
    }
}
```

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.h:
```diff
@@ -0,0 +1,101 @@
+    bool m_bCustomDiff;
+
+    // fadeout animation fields
+    float *m_pFlAlpha;
+    char *m_pSzAnimationName;
```
The type part of hungarian notation shouldn't ever be capitalized.
```suggestion
    float *m_pflAlpha;
    char *m_pszAnimationName;
```

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.h:
```diff
@@ -0,0 +1,101 @@
+
+    SpeedometerColorize_t GetColorizeType() { return m_eColorizeType; }
+    void SetColorizeType(SpeedometerColorize_t type) { m_eColorizeType = type; Reset(); }
+    void SetColorizeType(int type) { SetColorizeType(SpeedometerColorize_t(type)); }
```
You should probably bounds check this `type` param with the enum, by making a `SPEEDOMETER_COLORIZE_FIRST` and `SPEEDOMETER_COLORIZE_LAST`. The [network packet reading stuff](https://github.com/momentum-mod/game/blob/develop/mp/src/game/shared/momentum/mom_ghostdefs.h#L359) did this recently, but it's generally good practice to do this when converting int -> enum.

Gocnak commented on mp/src/game/client/momentum/ui/SettingsPanel/HudSettingsPage.cpp:
```diff
@@ -3,20 +3,47 @@
 {
-    m_pSpeedometerUnits = new CvarComboBox(this, "SpeedoUnits", "mom_hud_speedometer_units");
+    m_pSpeedometerGameType = new ComboBox(this, "SpeedoGameType", GAMEMODE_COUNT - 1, false);
+    for (auto i = 1; i < sizeof(g_szGameModes) / sizeof(*g_szGameModes); i++)
```
```suggestion
    for (auto i = 1; i < GAMEMODE_COUNT; i++)
```

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,200 @@
+    if (pGamemodeKV) // game mode is known
+    {
+        // set speedometer label setups
+        for (KeyValues *pSpeedoItem = pGamemodeKV->GetFirstSubKey(); pSpeedoItem != nullptr;
+             pSpeedoItem = pSpeedoItem->GetNextKey())
```
```suggestion
        FOR_EACH_SUBKEY(pGamemodeKV, pSpeedoItem)
```

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,200 @@
+            if (!pLabel || !pSpeedoKV)
+                continue;
+
+            pSpeedoKV->Clear();
+            pSpeedoKV->SetBool("visible", pLabel->IsVisible());
+            pSpeedoKV->SetInt("colorize", pLabel->GetColorizeType());
+            RangeList *pRangeList = pLabel->GetRangeList();
+            if (pRangeList)
+            {
+                KeyValues *pRangesKV = pSpeedoKV->FindKey("ranges", true);
+                pRangesKV->Clear();
+                char tmp[BUFSIZELOCL];
+                for (auto i = 0; i < pRangeList->Count(); i++)
+                {
+                    Range_t *range = &(*pRangeList)[i];
+                    Q_snprintf(tmp, BUFSIZELOCL, "%i", i + 1);
+                    KeyValues *rangeKV = pRangesKV->FindKey(tmp, true);
+                    rangeKV->SetInt("min", range->min);
+                    rangeKV->SetInt("max", range->max);
+                    rangeKV->SetColor("color", range->color);
+                }
+            }
+            pSpeedoKV->SetInt("units", pLabel->GetUnitType());
```
This could actually move back into the speedometer label under a `SaveToKV(KeyValues *pOut)` method.

This will keep all the label's settings in the label's class, meaning you could also get rid of `GetRangeList`.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,200 @@
+            if (!pLabel || !pSpeedoKV)
+                continue;
+
+            pLabel->SetVisible(pSpeedoKV->GetBool("visible", false));
+            if (!pLabel->IsVisible())
+                continue;
+
+            int colorize = pSpeedoKV->GetInt("colorize", 0);
+            if (colorize == SPEEDOMETER_COLORIZE_RANGE)
+            {
+                KeyValues *pRangesKV = pSpeedoKV->FindKey("ranges");
+                if (!pRangesKV)
+                    continue;
+
+                RangeList *m_pRangeList = new RangeList();
+                for (KeyValues *pRangeItem = pRangesKV->GetFirstSubKey(); pRangeItem != nullptr;
+                     pRangeItem = pRangeItem->GetNextKey())
+                {
+                    Range_t range;
+                    range.min = pRangeItem->GetInt("min", 0);
+                    range.max = pRangeItem->GetInt("max", 0);
+                    range.color = pRangeItem->GetColor("color");
+                    m_pRangeList->AddToTail(range);
+                }
+                pLabel->SetRangeList(m_pRangeList);
+            }
+            pLabel->SetColorizeType(colorize);
+            pLabel->SetUnitType(pSpeedoKV->GetInt("units", 0));
```
Likewise here, this can be under a `LoadFromKV(KeyValues *pIn)` method for the label class. This would get rid of the need to store the RangeList as a pointer (can be a class member object instead).

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,200 @@
+        }
+
+        // get speedometer label setups
+        for (KeyValues *pSpeedoItem = pGamemodeKV->GetFirstSubKey(); pSpeedoItem != nullptr;
+             pSpeedoItem = pSpeedoItem->GetNextKey())
```
```suggestion
        FOR_EACH_SUBKEY(pGamemodeKV, pSpeedoItem)
```

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,200 @@
+{
+    if (args.ArgC() > 1)
+    {
+        g_pSpeedometerData->LoadGamemodeData(g_pGameModeSystem->GetGameMode(Q_atoi(args[1]))->GetType());
```
You should probably check the outcome of `Q_atoi(args[1])` and see if it's a valid value, otherwise we'd get a nasty crash here. Do this for both load and save.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+{
+    BaseClass::PerformLayout();
+
+    // keep centered in parent
```
Labels currently center pretty nicely internally. Should this code only run if there's a comparison label being drawn?

braem replied:
They do unless they are auto wide, and it needs to be auto wide in order to pin the comparison to the side of it. I suppose I could force auto wide on/off depending on if comparison will be drawn instead.

Gocnak replied:
Eh, don't worry about it

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+    {
+        m_flDiff = fabs(m_flCurrentValue) - fabs(m_flPastValue);
+    }
+    SetFgColor(
+        MomUtil::GetColorFromVariation(m_flDiff, COLORIZE_DEADZONE, m_NormalColor, m_IncreaseColor, m_DecreaseColor));
```
Can be same line

braem replied:
Stupid visual studio.. must've happened when I copy pasted the code into it's own function

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+void SpeedometerLabel::ColorizeComparisonSeparate()
+{
+    SetFgColor(m_NormalColor);
+    if (m_bDrawComparison)
+    {
```
Early return if false.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.h:
```diff
@@ -0,0 +1,101 @@
+// for ranged based coloring
+struct Range_t
+{
+    int min;
+    int max;
```
You should probably mention in a comment that these are inclusive.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+}
+
+// creates a separate label for the comparison & pins it to this
+void SpeedometerLabel::CreateComparisonLabel()
+{
+    char name[BUFSIZELOCL];
+    Q_snprintf(name, sizeof(name), "%sComparison", GetName());
+    m_pComparisonLabel = new Label(GetParent(), name, "");
+    m_pComparisonLabel->SetName(name);
+    m_pComparisonLabel->SetPos(0, 0);
+    m_pComparisonLabel->PinToSibling(GetName(), PIN_TOPLEFT, PIN_TOPRIGHT);
+    m_pComparisonLabel->SetFont(GetFont());
+    m_pComparisonLabel->SetAutoTall(true);
+    m_pComparisonLabel->SetAutoWide(true);
+    m_pComparisonLabel->SetContentAlignment(a_center);
+    m_pComparisonLabel->SetPinCorner(PIN_TOPLEFT, 0, 0);
+}
```
This can actually move to the constructor, it's small enough and is actually constructing vital things for the class.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+    : Label(parent, panelName, ""), m_eColorizeType(colorizeType), m_pFlAlpha(nullptr),
+      m_pSzAnimationName(nullptr), m_bSupportsEnergyUnits(false), m_bDoneFading(false),
+      m_eUnitType(SPEEDOMETER_UNITS_UPS), m_pRangeList(nullptr), m_bDrawComparison(true), 
+      m_bSupportsSeparateComparison(true), m_cvarGravity("sv_gravity")
```
Instead of using a cvar ref for gravity, do:
```
#include "movevars_shared.h"

. . . 

using namespace vgui;

extern ConVar sv_gravity;
```

And use `sv_gravity` directly.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,285 @@
+    }
+    else
+    {
+        m_flDiff = fabs(m_flCurrentValue) - fabs(m_flPastValue);
```
You can take just the `fabs` of the result:
```
m_flCurrentValue = -30
m_flPastValue = -20

| -30 | - | -20 | = 30 - 20 = 10
| -30 - -20 | = | -10 | = 10
```

braem replied:
I don't think we can, as doing that will make `m_flDiff` always positive, meaning that it will never be colored red for a loss of speed. 

Looks like we don't need to get the absolute value here, as the current and past value will always be positive since speed is a magnitude. The only exception being energy units, which also doesn't have the `fabs` calls. I removed them.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -1,570 +1,185 @@
         {
-            DeactivateLabel(m_pAbsSpeedoLabel);
+            pLabel->SetPos(pLabel->GetXPos(), iHeightAcc);
+            iHeightAcc += surface()->GetFontTall(pLabel->GetFont());
```
Right, I changed it to use `GetTall()` instead. Also removed the `ISurface` include as that's all that was for.

braem commented:
Finished the requested changes. Still investigating how `PerformLayout` could be done better in the speedo label. Simply enabling/disabling auto wide didn't work as I'd hoped.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -1,570 +1,183 @@
-            }
-            int offsetXPos = 0 - ((GetWide() - combinedLength) / 2);
+    float absVel = static_cast<float>(velocity.Length());
+    float horizVel = static_cast<float>(Vector(velocity.x, velocity.y, 0).Length());
```
Length returns a vec_t which is a typedef to float

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.h:
```diff
@@ -0,0 +1,57 @@
+#include "hudelement.h"
+
+#include <vgui_controls/EditablePanel.h>
+#include "mom_player_shared.h"
```
Can this also move to the cpp?

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.h:
```diff
@@ -0,0 +1,57 @@
+    class Label;
+}
+
+typedef CUtlVector <SpeedometerLabel *> SpeedoLabelList;
```
You can keep the type brackets close to the class, this makes it look like there's three things here:

`CUtlVector<SpeedometerLabel*>`

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.h:
```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include "hud_speedometer_label.h"
```
Can you forward declare this as well?

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.cpp:
```diff
@@ -0,0 +1,165 @@
+        int gamemodeIndex = Q_atoi(args[1]);
+        if (gamemodeIndex >= 0 && gamemodeIndex < GAMEMODE_COUNT)
+        {
+            g_pSpeedometerData->SaveGamemodeData(g_pGameModeSystem->GetGameMode(Q_atoi(args[1]))->GetType());
```
Use gamemodeIndex here

Gocnak replied:
On second thought you could just cast the index to the type

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.h:
```diff
@@ -0,0 +1,57 @@
+    void Reset() OVERRIDE;
+    void FireGameEvent(IGameEvent *pEvent) OVERRIDE;
+    void ApplySchemeSettings(vgui::IScheme *pScheme) OVERRIDE;
+    void PerformLayout() OVERRIDE;
```
OVERRIDE -> override

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_data.h:
```diff
@@ -0,0 +1,31 @@
+#include "mom_shareddefs.h"
+#include "mom_system_gamemode.h"
+
+#include "hud_speedometer.h"
```
Forward declare and move this to the cpp

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.h:
```diff
@@ -0,0 +1,100 @@
+
+    vgui::Label *m_pComparisonLabel;
+
+    RangeList m_pRangeList;
```
This is no longer a pointer so you can use `m_vecRangeList`

braem commented:
Implemented the requested changes. Did another pass of the imports & managed to move most of them to the source files by using forward declarations of needed classes. Moved the `LoadGamemodeData` functions to the source file to avoid needing to import `mom_system_gamemode.h` in the header.

Also found a bug where reloading controls forced all speedos to be visible. This is makes sense as `Panel.cpp` sets the visibility to `true` regardless of whether it's defined in the resource file, in both `ApplySchemeSettings` and `ApplySettings`. I'm guessing the best way to handle this is move away from using the baseclass's visibility, instead having a local variable that determines visibility as set in the VDF file. Or just reverse the change a la
`
bool bIsVisible = IsVisible();
BaseClass::ApplySchemeSettings(pScheme);
SetVisible(bIsVisible);
`
Figured I'd ask your opinion before this. Can always make a separate PR for that when this gets merged. I know there will be a few more (adding ramp enter/exit speed & vert speedo). I also want to throw another PR up to stop stage enter speed from showing multiple times on the same zone, as in when u fail it shows `0` enter speed.

braem commented:
Added that `OnReloadControls()` override that reloads the file as suggested.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,329 @@
+
+void SpeedometerLabel::SetFadeOutAnimation(char *animationName, float *animationAlpha)
+{
+    m_pszAnimationName = animationName;
```
You should `Q_strncpy` into a char array instead of just storing a pointer, and the check to see if it's valid is just checking if the first char is `\0`.

Gocnak commented on mp/game/momentum/resource/momentum_english_ref.res:
```diff
@@ -273,22 +273,24 @@
+        "MOM_Settings_Speedometer_Type_Horiz" "Horizontal Speedometer"
+        "MOM_Settings_Speedometer_Type_LastJump" "Last Jump Velocity"
+        "MOM_Settings_Speedometer_Type_StageEnterExit" "Stage Enter/Exit Velocity"
+        "MOM_Settings_Speedometer_Show" "Display this speedometer."
```
Should this have a period? I didn't see any pictures of any UI so I don't know the context of this string

braem replied:
Fits with the key presses & strafe sync stuff but not the timer.
![image](https://user-images.githubusercontent.com/9014762/83776454-0b3d3d00-a63d-11ea-84d8-a6d0bdbd98f8.png)

Gocnak replied:
Ah alright, I'll clean this up in my options panel branch.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.h:
```diff
@@ -0,0 +1,58 @@
+
+#define SPEEDOMETER_MAX_LABELS 4
+
+namespace vgui 
+{
+    class Label;
+}
```
Is this needed? You don't seem to use any labels in this class, only SpeedometerLabel.

braem replied:
Another oops. Removed this.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,329 @@
+    pOut->SetBool("visible", IsVisible());
+    pOut->SetInt("colorize", GetColorizeType());
+    KeyValues *pRangesKV = pOut->FindKey("ranges", true);
+    pRangesKV->Clear();
```
Does this need to happen given you do `pOut->Clear()` above?

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp:
```diff
@@ -0,0 +1,329 @@
+    if (!IsVisible())
+        return;
+
+    int colorize = pIn->GetInt("colorize", 0);
```
```suggestion
    int colorize = pIn->GetInt("colorize", SPEEDOMETER_COLORIZE_NONE);
```

braem commented:
> 
> 
> As we wrap up review here, did you make sure to test in various maps (gamemodes) and spectating both online/replays? Just to catch any potential introduced bugs, if any.

I tested in replays and online spectating just now with false. There are a couple issues, though they exist in 0.8.4:
- last jump vel doesn't show in replays, as in 0.8.4. I see there is commented out code for that in `mom_replay_entity.cpp`.
- last jump vel doesn't show when spectating others, which makes sense in my head as they don't have the code to fire the jump event. Didn't test with anyone that's built this branch.
- timer wasn't starting when spectating other players. Not sure if that's a known limitation but I was getting that on 0.8.4 as well. Stage enter/exit works fine in replays as the timer starts there.
![image](https://user-images.githubusercontent.com/9014762/83785317-fe711700-a645-11ea-8a77-7ad8fec53a53.png)

Only other bug is related to the UI. When the settings are left open & player changes gamemode, it doesnt load the speedo settings into the hud settings page until it's tabbed off. Is there some `OnSomething` override that fires when focus is lost via escape key? Figured it wasn't so important as you are changing a large portion of the UI anyhow.

Gocnak commented:
Yeah with the new settings panel it's modal, meaning you can't go back to the gameplay without closing it (in normal circumstances), so that shouldn't be a problem.
